클린아키텍쳐 레이어 분리

hooni0918 opened the pull request:
코드가 좀 복잡한데 고민을 많이하고 시간이 많이 없어서 코드가 더러운점 양해바랍니다.
--

코드 구조

```mermaid
graph TD
    subgraph "Presentation Layer"
        VC[ViewController]
        VM[ViewModel]
        PM[Presentation Model]
    end

    subgraph "Domain Layer"
        IF[Interface]
        UC[UseCase]
        EN[Entity]
    end

    subgraph "Data Layer"
        DTO[DTOs]
        MAP[Mappers]
        NET[Network]
        REPO[Repositories]
    end

    %% Dependencies
    VC --> VM
    VM --> UC
    VM --> PM
    PM --> EN
    UC --> IF
    IF --> EN
    REPO --> IF
    REPO --> NET
    REPO --> MAP
    MAP --> DTO
    MAP --> EN
    NET --> DTO

    %% Styling
    classDef presentation fill:#FFE4E1,stroke:#FF69B4
    classDef domain fill:#E0FFFF,stroke:#4169E1
    classDef data fill:#F0FFF0,stroke:#98FB98

    class VC,VM,PM presentation
    class IF,UC,EN domain
    class DTO,MAP,NET,REPO data
```


1. **Presentation Layer (UI Layer) :** 

사용자 인터페이스와 관련된 모든 요소를 처리

1-1  `MovieViewController`

- UI와 관련된 모든 요소들이 위치 ( 편의상 View를 따로 분리하지는 않았습니다.)
- 사용자 입력을 처리하고 화면 표시
- 화면의 생명주기 + UI 이벤트를 관리

1-2 `ViewModel Layer`

- Presentation과 Domain 레이어 사이에서 중개
- 비즈니스 로직을 처리

1-3 `MoviePresentationModel`, `MoviePresentationMapper`

- UI용 표시될 데이터 구조 정의
- 도메인 모델을 UI 형태로 변환 (날짜 숫자 포맷팅)

1. **Domain Layer (Business Layer)**
- 목적: 비즈니스 규칙과 로직을 포함하는 코어한 레이어

2-1 `Interface`: `MovieRepository` (protocol)

- 데이터 계층과의 통신을 위한 추상화 인터페이스
- 구체적인 구현에 의존하지않음
- 도메인 모델로 통신

2-2 `UseCase`: `FetchMoviesUseCase`

- 구체적인 비즈니스 로직 구현
- 단일 책임 원칙을 따름
- Repository를 통해 데이터 접근
- 도메인 모델을 사용하여 작업

2-3 `Entity`: `Movie`

- 핵심 비즈니스 규칙을 포함한 도메인 모델
- 순수한 **비즈니스 로직**만 포함
- 외부 의존성이 없음

1. **Data Layer**
- 목적: 외부 데이터 소스와의 통신 및 데이터 변환을 처리

`DTOs`: BoxOfficeDTO

`Mappers`:BoxOfficeMapper

- DTO를 도메인 모델로 변환

`Network`:`BoxOfficeNetworkService`, `BoxOfficeAPI`

- 네트워크 요청/응답 처리 등

`Repositories`: `DefaultMovieRepository`

- Repository 인터페이스
- 네트워크 서비스와 매퍼를 조합하여 데이터 처리

**레이어 분리 기준**:

1. 의존성 방향: 안쪽 레이어(Domain)는 바깥쪽 레이어(Data, Presentation)를 모르도록 설계
2. 관심사 분리: 각 레이어는 자신의 고유한 책임만 가짐
3. 추상화 수준: Domain Layer는 가장 추상적이고( "무엇을" 하는지만 정의하고 "어떻게" 하는지는 감추는 것) ,  Data와 Presentation Layer는 구체적인 구현을 포함해서 구현
4. 모듈성: 각 레이어는 독립적으로 수정/교체 가능하게 구현


무튼 제 생각은 이렇습니다..
개인단위 혹은 작은 사이드 단위에서 클린아키텍쳐를 처음부터 도입하는게 정말 굳이? 싶다 라는것도 제 결론이네요ㅋㅋㅋ
덕분에 미루고 미뤗던 클린아키 개념을 잡고가는것 같아 다행입니다.
코드에 대한 질문 환영입니다

hooni0918 committed:
Add new Swift files related to movie data handling and networking. Organize files into DTOs, Mappers, Network, Repositories, Data, Entity, InterFace, UseCase, and Domain groups for better project structure.

thingineeer commented on Buttwink/Presentation/Movie/Domain/UseCase/FetchMoviesUseCase.swift:
```diff
@@ -0,0 +1,24 @@
+    func execute(date: Date) async throws -> [Movie]
+}
+
+final class DefaultFetchMoviesUseCase: FetchMoviesUseCase {
+    private let movieRepository: MovieRepository
+    
+    init(movieRepository: MovieRepository) {
+        self.movieRepository = movieRepository
+    }
+    
+    func execute(date: Date) async throws -> [Movie] {
+        return try await movieRepository.fetchDailyBoxOffice(date: date)
+    }
```
질문 있어요!

`DefaultFetchMoviesUseCase` 는 `MovieRepository`(protocol)을 의존하고 있고

`DefaultMovieRepository` 가 `MovieRepository`(protocol) 을 구현하는 부분인데

이 부분이 아래 사진에서 두개의 화살표에 해당하는 부분이 맞을까요?

<img width="612" alt="image" src="https://github.com/user-attachments/assets/3c3a6ca0-78c9-49c5-97b5-3d12ba711c69" />

hooni0918 replied:
정답입니다~!

thingineeer commented on Buttwink/Presentation/Movie/Presentation/ViewController/MovieViewController.swift:
```diff
@@ -36,8 +36,27 @@ final class MovieViewController: UIViewController {
-    // MARK: - Initialization
-    init(viewModel: MovieViewModelType = MovieViewModel()) {
+    convenience init() {
+        let networkService = DefaultBoxOfficeNetworkService()
+        let boxOfficeMapper = DefaultBoxOfficeMapper()
+        let movieRepository = DefaultMovieRepository(
+            networkService: networkService,
+            mapper: boxOfficeMapper
+        )
+        let fetchMoviesUseCase = DefaultFetchMoviesUseCase(
+            movieRepository: movieRepository
+        )
+        let presentationMapper = DefaultMoviePresentationMapper()
+        
+        let viewModel = MovieViewModel(
+            fetchMoviesUseCase: fetchMoviesUseCase,
+            moviePresentationMapper: presentationMapper
+        )
+        
+        self.init(viewModel: viewModel)
```
질문 있습니다. 
convenience init 부분에서 따로 의존성 객체를 만들어서 외부 주입을 시켜주셨는데, 원래는 이 부분을 DIContainer를 통해 해결하는 게 맞죠?

hooni0918 replied:
정확합니다

thingineeer commented on Buttwink/Presentation/Movie/Presentation/ViewModel/MovieViewModel.swift:
```diff
@@ -0,0 +1,95 @@
+                    
+                    let presentationMovies = self.moviePresentationMapper.transform(domainMovies)
+                    
+                    await MainActor.run {
+                        self.movies.accept(presentationMovies)
+                        self.isLoading.accept(false)
+                        observer.onNext(presentationMovies)
+                        observer.onCompleted()
+                    }
+                } catch {
+                    await MainActor.run {
+                        self.error.accept(error)
+                        self.isLoading.accept(false)
+                        observer.onError(error)
+                    }
+                }
```
공부를 하다가, 어쨌든 MainActor는 ViewModel이나 UIViewController에서 많이 사용되는 걸로 알고 있습니다.
@MainActor든, MainActor.run이든 !

저 같은 경우는 바인딩 하는 부분, 즉 ViewController에서 사용 했었는데, ViewModel에서 사용하는 것과 ViewController에서 사용하는 것의 기준이 있으실까요?

hooni0918 replied:
지금 다시 생각해보니 해당 코드는 Rx를 사용해서 `observe(on:)` 을 사용하는게 맞는 코드이겟네요 
Viewcontroller가 아닌  뷰 모델에서는 데이터를 바인딩할때 사용할 필요없이 오직 뷰컨에서 UI관련 로직에서만 사용하는게 맞는 코드라 생각합니다.
메인액터를 남발할경우, 직렬화시키는 액터의 특성상 불필요한 리소스를 너무많이 사용할것이기에 UI 업데이트 + 뷰컨이나 뷰 정도에서만 사용하는것이 옳을듯합니다.

thingineeer reviewed (commented):
따로 Mapper로 빼놓는 것도 좋게 봤고, 머메이드를 활용해서 PR 날려주신 거 잘 봤습니다.
코드 보면서 많이 배웠고 저도 열심히 공부하겠습니다.

GodSamble commented on Buttwink/Presentation/Movie/Data/Repositories/DefaultMovieRepository.swift:
```diff
@@ -0,0 +1,23 @@
+    private let networkService: BoxOfficeNetworkService
+    private let mapper: BoxOfficeMapper
+    
+    init(networkService: BoxOfficeNetworkService, mapper: BoxOfficeMapper) {
```
매퍼따로 빼서 관리한 게 인상적이고,
파라미터 인자로 '서비스'하고 '매퍼'를 넣는거 좋네요.

hooni0918 replied:
영광입니다

GodSamble commented:
<클아키의 정석> 출판해도 되겠네요 너무 FM 이네요 👍
SOLID 의거한 클린아키였어서 보는 내내 마음이 편했어요 :)

GodSamble reviewed (commented):
<클아키의 정석> 출판해도 되겠네요 너무 FM 이네요 👍
SOLID 의거한 클린아키였어서 보는 내내 마음이 편했어요 :)

hooni0918 committed:
Merge remote-tracking branch 'GodSamble/main'

hooni0918 committed:
Refactor BoxOffice and MovieViewModel structures

- Updated BoxOfficeResponse structure in BoxOffice.swift
- Modified fetchDailyBoxOffice method return type to [Movie] in ServiceType.swift
- Changed class declaration to class MovieViewModel in MovieViewModel.swift
